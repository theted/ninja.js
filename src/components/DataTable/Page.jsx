import React from 'react';
import PropTypes from 'prop-types';

const Page = ({ pageNumber, currentPageNumber, onChange }) => {
  const isActivePage = () => currentPageNumber === pageNumber;

  const renderedPageNumber = () => pageNumber + 1;

  const click = (event) => {
    event.preventDefault();
    onChange(pageNumber);
  };

  const className = `page-link${isActivePage()}` ? ' button-outline' : '';

  return (
    <li className="page-item mr-1">
      <button className={className} onClick={click} type="button">
        {renderedPageNumber()}
      </button>
    </li>
  );
};

Page.propTypes = {
  pageNumber: PropTypes.number,
  currentPageNumber: PropTypes.number,
  onChange: PropTypes.func,
};

Page.defaultProps = {
  pageNumber: 0,
  currentPageNumber: 0,
  onChange: () => {},
};

export default Page;
