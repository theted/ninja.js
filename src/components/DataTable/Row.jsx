import React from 'react';
import PropTypes from 'prop-types';

const Row = ({ row }) => {
  return (
    <tr>
      <td>
        <a href={row.edit_path}>
          {row.name1}
        </a>
        <br />
        <small>{row.email}</small>
      </td>
    </tr>
  );
};

Row.propTypes = {
  row: PropTypes.arrayOf(PropTypes.object).isRequired,
}

export default Row;
