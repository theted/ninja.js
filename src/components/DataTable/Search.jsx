import React from 'react';

const Search = ({ onSearch }) => (
  <div className="p-b-1">
    <input
      type="search"
      className="form-control"
      placeholder="SÃ¸g brugere"
      onChange={onSearch}
    />
  </div>
);

export default Search;
