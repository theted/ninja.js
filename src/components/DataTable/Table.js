import React from 'react';

const Table = ({ body, layout = 'test' }) => {
  return (
    <table>
      <tbody>{body}</tbody>
    </table>
  );
};

export default Table;
