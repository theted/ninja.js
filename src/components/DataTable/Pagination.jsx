import React from 'react';
import PropTypes from 'prop-types';

import Page from './Page';

const Pagination = ({ currentPageNumber, totalNumberOfPages, onChange }) => {
  const pages = Array.from(Array(totalNumberOfPages).keys()).map((pageNumber) => (
    <Page
      key={pageNumber}
      currentPageNumber={currentPageNumber}
      pageNumber={pageNumber}
      onChange={onChange}
    />
  ));

  if (pages.length <= 1) {
    return null;
  }
  return <ul className="pagination">{pages}</ul>;
};

Pagination.propTypes = {
  currentPageNumber: PropTypes.number.isRequired,
  totalNumberOfPages: PropTypes.number.isRequired,
  onChange: PropTypes.func.isRequired,
};

export default Pagination;
