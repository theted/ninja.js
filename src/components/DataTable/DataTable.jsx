import React from 'react'

import Pagination from './Pagination'
import Row from './Row'
import Search from './Search'

class DataTable extends React.Component {
  state = {
    rows: this.props.rows,
    currentPageNumber: 0,
    totalNumberOfPages: this.calculateTotalNumberOfPages(this.props.rows)
  }

  static defaultProps = {
    rowsPerPage: 40
  }

  calculateTotalNumberOfPages(rows) {
    const { rowsPerPage } = this.props
    if (rowsPerPage == 0) return 0
    return Math.ceil(rows.length / rowsPerPage)
  }

  search(event) {
    const { rows } = this.props
    const text = event.target.value
    let rowsFound = rows

    if (text) {
      rowsFound = rows.filter((row) => {
        return row.name1.toLowerCase().search(text.toLowerCase()) > -1 ||
         (row.email && row.email.toLowerCase().search(text.toLowerCase()) > -1)
      })
    }

    this.setState({
      rows: rowsFound,
      currentPageNumber: 0,
      totalNumberOfPages: this.calculateTotalNumberOfPages(rowsFound)
    })
  }

  changeToPageNumber(pageNumber) {
    this.setState({ currentPageNumber: pageNumber })
  }

  rowsInPageNumber(pageNumber) {
    const { rowsPerPage } = this.props
    const startIndex = pageNumber * rowsPerPage
    return [startIndex, startIndex + rowsPerPage]
  }

  render() {
    const { rows, currentPageNumber, totalNumberOfPages } = this.state
    const rowsToRender = rows
      .map(row => <Row key={row.per_id} row={row} />)
      .slice(...this.rowsInPageNumber(currentPageNumber))

    return(
      <div>
        <Search onSearch={this.search.bind(this)} />
        <table>
          <tbody>
            { rowsToRender }
          </tbody>
        </table>
        <Pagination
          currentPageNumber={currentPageNumber}
          totalNumberOfPages={totalNumberOfPages}
          onChange={this.changeToPageNumber.bind(this)} />
      </div>
    )
  }
}

export default DataTable
