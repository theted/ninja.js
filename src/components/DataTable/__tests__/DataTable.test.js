import React from 'react';
import { mount } from 'enzyme'; // Lifecycle hooks such as useEffect aren’t yet supported in shallow render (those hooks don’t get called)
import DataTable from '..';
import users from '../../../data/users.json';

const props = {
  rows: users.slice(0, 5),
  locale: 'da',
  rowsPerPage: 5,
};
describe('DataTable Component', () => {
  it('renders without crashing', () => {
    mount(<DataTable {...props} />);
  });

  it('matches snapshot', () => {
    const component = mount(<DataTable {...props} />);
    component.update();
    expect(component).toMatchSnapshot();
  });

  it('renders 5 rows', () => {
    const wrapper = mount(<DataTable {...props} />);

    expect(wrapper.find('tr').length).toBe(5);
  });

  it('handles perPage prop', () => {
    const wrapper = mount(<DataTable {...props} rowsPerPage={2} />);

    expect(wrapper.find('tr').length).toBe(2);
  });

  it('filters rows based on input', () => {
    const wrapper = mount(<DataTable {...props} />);

    wrapper.find('input').simulate('change', { target: { value: 'k' } });

    expect(wrapper.find('tr').length).toBe(2);
  });

  it('correctly clears filters', () => {
    const wrapper = mount(<DataTable {...props} />);

    wrapper
      .find('input')
      .simulate('change', { target: { value: 'k' } })
      .simulate('change', { target: { value: '' } });

    expect(wrapper.find('tr').length).toBe(5);
  });

  it('return correct number of pages', () => {
    const wrapper = mount(<DataTable {...props} rowsPerPage={3} />);
    expect(wrapper.find('li').length).toBe(2);
  });
});
