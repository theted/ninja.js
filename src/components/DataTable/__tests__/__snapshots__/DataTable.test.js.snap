// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable Component matches snapshot 1`] = `
<DataTable
  locale="da"
  rows={
    Array [
      Object {
        "edit_path": "http://google.com",
        "email": "MadsLKlausen@jourrapide.com",
        "name1": "Mads L. Klausen",
        "per_id": 1,
      },
      Object {
        "edit_path": "http://google.com",
        "email": "AlfredKKrogh@armyspy.com",
        "name1": "Alfred K. Krogh",
        "per_id": 2,
      },
      Object {
        "edit_path": "http://google.com",
        "email": "SilasLBertelsen@armyspy.com",
        "name1": "Silas L. Bertelsen",
        "per_id": 3,
      },
      Object {
        "edit_path": "http://google.com",
        "email": "MiaAJohnsen@dayrep.com",
        "name1": "Mia A. Johnsen",
        "per_id": 4,
      },
      Object {
        "edit_path": "http://google.com",
        "email": "AlfredSSchou@jourrapide.com",
        "name1": "Alfred S. Schou",
        "per_id": 5,
      },
    ]
  }
  rowsPerPage={5}
>
  <div>
    <Search
      onSearch={[Function]}
    >
      <div
        className="p-b-1"
      >
        <input
          className="form-control"
          onChange={[Function]}
          placeholder="SÃ¸g brugere"
          type="search"
        />
      </div>
    </Search>
    <Table
      body={
        Array [
          <Row
            row={
              Object {
                "edit_path": "http://google.com",
                "email": "MadsLKlausen@jourrapide.com",
                "name1": "Mads L. Klausen",
                "per_id": 1,
              }
            }
          />,
          <Row
            row={
              Object {
                "edit_path": "http://google.com",
                "email": "AlfredKKrogh@armyspy.com",
                "name1": "Alfred K. Krogh",
                "per_id": 2,
              }
            }
          />,
          <Row
            row={
              Object {
                "edit_path": "http://google.com",
                "email": "SilasLBertelsen@armyspy.com",
                "name1": "Silas L. Bertelsen",
                "per_id": 3,
              }
            }
          />,
          <Row
            row={
              Object {
                "edit_path": "http://google.com",
                "email": "MiaAJohnsen@dayrep.com",
                "name1": "Mia A. Johnsen",
                "per_id": 4,
              }
            }
          />,
          <Row
            row={
              Object {
                "edit_path": "http://google.com",
                "email": "AlfredSSchou@jourrapide.com",
                "name1": "Alfred S. Schou",
                "per_id": 5,
              }
            }
          />,
        ]
      }
    >
      <table>
        <tbody>
          <Row
            key="1"
            row={
              Object {
                "edit_path": "http://google.com",
                "email": "MadsLKlausen@jourrapide.com",
                "name1": "Mads L. Klausen",
                "per_id": 1,
              }
            }
          >
            <tr>
              <td>
                <a
                  href="http://google.com"
                >
                  Mads L. Klausen
                </a>
                <br />
                <small>
                  MadsLKlausen@jourrapide.com
                </small>
              </td>
            </tr>
          </Row>
          <Row
            key="2"
            row={
              Object {
                "edit_path": "http://google.com",
                "email": "AlfredKKrogh@armyspy.com",
                "name1": "Alfred K. Krogh",
                "per_id": 2,
              }
            }
          >
            <tr>
              <td>
                <a
                  href="http://google.com"
                >
                  Alfred K. Krogh
                </a>
                <br />
                <small>
                  AlfredKKrogh@armyspy.com
                </small>
              </td>
            </tr>
          </Row>
          <Row
            key="3"
            row={
              Object {
                "edit_path": "http://google.com",
                "email": "SilasLBertelsen@armyspy.com",
                "name1": "Silas L. Bertelsen",
                "per_id": 3,
              }
            }
          >
            <tr>
              <td>
                <a
                  href="http://google.com"
                >
                  Silas L. Bertelsen
                </a>
                <br />
                <small>
                  SilasLBertelsen@armyspy.com
                </small>
              </td>
            </tr>
          </Row>
          <Row
            key="4"
            row={
              Object {
                "edit_path": "http://google.com",
                "email": "MiaAJohnsen@dayrep.com",
                "name1": "Mia A. Johnsen",
                "per_id": 4,
              }
            }
          >
            <tr>
              <td>
                <a
                  href="http://google.com"
                >
                  Mia A. Johnsen
                </a>
                <br />
                <small>
                  MiaAJohnsen@dayrep.com
                </small>
              </td>
            </tr>
          </Row>
          <Row
            key="5"
            row={
              Object {
                "edit_path": "http://google.com",
                "email": "AlfredSSchou@jourrapide.com",
                "name1": "Alfred S. Schou",
                "per_id": 5,
              }
            }
          >
            <tr>
              <td>
                <a
                  href="http://google.com"
                >
                  Alfred S. Schou
                </a>
                <br />
                <small>
                  AlfredSSchou@jourrapide.com
                </small>
              </td>
            </tr>
          </Row>
        </tbody>
      </table>
    </Table>
    <Pagination
      currentPageNumber={0}
      onChange={[Function]}
      totalNumberOfPages={1}
    />
  </div>
</DataTable>
`;
